EKF2. Use existing sensor priority to spawn multi-ekf instances.

antonio-sc66 opened the pull request:
### Solved Problem
EKF2.cpp isn't taking into account sensor priorities. Instances are spawned following a basic 0 to maximum number of sensors approach.

Uninitialized or disabled sensors are being used, which results in spawned EKF instances with wrong sensors, this takes up more memory (I think so, might be wrong) and CPU.

Fixes #21225

### Solution
- Read priorities from parameters
- Discard sensors that are uninitialized or disabled
- Update the calculation of the maximum number of EKF instances that should be spawned
- Only try to spawn the correct number of EKF instances

### Changelog Entry
For release notes:
```
Feature/Bugfix XYZ
New parameter: XYZ_Z
Documentation: Need to clarify page ... / done, read docs.px4.io/...
```

### Test coverage
- Simulation/hardware testing logs: https://review.px4.io/plot_app?log=d1aae94d-c9f4-4605-b2a3-2a7e5fedbee5
- Tested in SITL

### Test context
I enabled some Debug prints as Info to get a better understanding of how this new approach is working.

This flight was commanded both from the shell and from QGroundControl.

`make px4_sitl gazebo` flight output (commit: ded751e3ae) :

```
______  __   __    ___ 
| ___ \ \ \ / /   /   |
| |_/ /  \ V /   / /| |
|  __/   /   \  / /_| |
| |     / /^\ \ \___  |
\_|     \/   \/     |_/

px4 starting.

INFO  [px4] startup script: /bin/sh etc/init.d-posix/rcS 0
INFO  [init] found model autostart file as SYS_AUTOSTART=10015
INFO  [param] selected parameter default file parameters.bson
INFO  [param] importing from 'parameters.bson'
INFO  [parameters] BSON document size 404 bytes, decoded 404 bytes (INT32:14, FLOAT:6)
INFO  [param] selected parameter backup file parameters_backup.bson
INFO  [dataman] data manager file './dataman' size is 7866640 bytes
etc/init.d-posix/rcS: 39: [: Illegal number: 
INFO  [init] PX4_SIM_HOSTNAME: localhost
INFO  [simulator_mavlink] Waiting for simulator to accept connection on TCP port 4560
INFO  [simulator_mavlink] Simulator connected on TCP port 4560.
INFO  [lockstep_scheduler] setting initial absolute time to 3352000 us
INFO  [commander] LED: open /dev/led0 failed (22)
WARN  [health_and_arming_checks] Preflight Fail: ekf2 missing data
INFO  [ekf2] Initialized MAGs (positive priority): 2
INFO  [ekf2] MAGs priorities - i: 0  MAG1  Priority: 50
INFO  [ekf2] MAGs priorities - i: 1  MAG0  Priority: 50
INFO  [ekf2] MAGs priorities - i: 2  MAG3  Priority: -1
INFO  [ekf2] MAGs priorities - i: 3  MAG2  Priority: -1
INFO  [ekf2] Initialized IMUs (positive priority): 1
INFO  [ekf2] IMUs priorities - i: 0  ACC0  Priority: 50
INFO  [ekf2] IMUs priorities - i: 1  ACC3  Priority: -1
INFO  [ekf2] IMUs priorities - i: 2  ACC2  Priority: -1
INFO  [ekf2] IMUs priorities - i: 3  ACC1  Priority: -1
INFO  [ekf2] all valid EKF instances have been started. Started instances: 1
INFO  [mavlink] mode: Normal, data rate: 4000000 B/s on udp port 18570 remote port 14550
Gazebo multi-robot simulator, version 11.8.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.1.42
INFO  [mavlink] mode: Onboard, data rate: 4000000 B/s on udp port 14580 remote port 14540
INFO  [mavlink] mode: Onboard, data rate: 4000 B/s on udp port 14280 remote port 14030
INFO  [mavlink] mode: Gimbal, data rate: 400000 B/s on udp port 13030 remote port 13280
INFO  [logger] logger started (mode=all)
INFO  [logger] Start file log (type: full)
INFO  [logger] [logger] ./log/2023-07-11/22_22_29.ulg	
INFO  [logger] Opened full log file: ./log/2023-07-11/22_22_29.ulg
INFO  [mavlink] MAVLink only on localhost (set param MAV_{i}_BROADCAST = 1 to enable network)
INFO  [mavlink] MAVLink only on localhost (set param MAV_{i}_BROADCAST = 1 to enable network)
INFO  [px4] Startup script returned successfully
pxh> INFO  [tone_alarm] home set
INFO  [tone_alarm] notify negative

pxh> 
pxh> INFO  [commander] Ready for takeoff!
pxh> commander takeoff
pxh> INFO  [commander] Armed by internal command	
INFO  [tone_alarm] arming warning
INFO  [navigator] Using minimum takeoff altitude: 2.50 m	

pxh> INFO  [commander] Takeoff detected	

pxh> ekf2 status
INFO  [ekf2] available instances: 1
INFO  [ekf2] 0: ACC: 1310988, GYRO: 1310988, MAG: 197388, healthy, test ratio: 0.0047039 (0.00000) *

ekf2:0 EKF dt: 0.0080s, attitude: 1, local position: 1, global position: 1
ekf2: ECL update: 1702 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: ECL full update: 1701 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: IMU message missed: 0 events
ekf2: vehicle_air_data messages missed: 0 events
ekf2: vehicle_gps_position messages missed: 0 events
ekf2: vehicle_magnetometer messages missed: 0 events
pxh> commander land
pxh> INFO  [commander] Landing at current position	
INFO  [commander] Landing detected	
INFO  [commander] Disarmed by landing	
INFO  [tone_alarm] notify neutral
INFO  [logger] closed logfile, bytes written: 6349087

pxh> shutdown
Exiting NOW.
```

### Testing information with current main

```
INFO  [ekf2] available instances: 1
INFO  [ekf2] 0: ACC: 1310988, GYRO: 1310988, MAG: 197388, healthy, test ratio: 0.0212503 (0.00000) *

ekf2:0 EKF dt: 0.0080s, attitude: 1, local position: 1, global position: 1
ekf2: ECL update: 5977 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: ECL full update: 5978 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: IMU message missed: 0 events
ekf2: vehicle_air_data messages missed: 0 events
ekf2: vehicle_gps_position messages missed: 0 events
ekf2: vehicle_magnetometer messages missed: 0 events

ekf2:1 EKF dt: 0.0100s, attitude: 0, local position: 0, global position: 0
ekf2: ECL update: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: ECL full update: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: IMU message missed: 0 events

ekf2:2 EKF dt: 0.0100s, attitude: 0, local position: 0, global position: 0
ekf2: ECL update: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: ECL full update: 0 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
ekf2: IMU message missed: 0 events
```


### More info
- *EKF2::task_spawn* method is too long. It should be splitted into smaller private methods or as functions in a utils file.
- Further testing has to be done and it should include real drones. I will try to do a test next week

antonio-sc66 committed:
EKF2. Use existing sensor priority to spawn multi-ekf instances.

- Sensor priorities (MAG and IMU) are taken into account. Higher priorities are spawned first.
- Disabled or uninitialized sensors are discarded and not spawned

Signed-off-by: Antonio Sanjurjo C <74329840+antonio-sc66@users.noreply.github.com>

antonio-sc66 commented:
Sorting has to be refactored so that it can be compiled to all targets

dagar commented on src/modules/ekf2/EKF2.cpp:
```diff
@@ -31,6 +31,8 @@
  *
  ****************************************************************************/
 
+#include <algorithm>
```
Unfortunately we don't have a full C++ standard library on the embedded side at the moment.

dagar commented:
Another idea I had was to move all of this runtime management to the estimator selector. Then we can handle dynamically starting/stopping instances based on current configuration, and things like the sensor priorities already need to be factored into selecting the primary.

antonio-sc66 commented:
That could make a lot of sense. It was strange for me to see that there's an EKF class and an EKFSelector class, but there's selector related stuff on the EKF class. It would involve a bigger refactor, that's for sure.

Another thing that I don't really understand is why _work queues_ are being selected purely based on the imu sensor number.

Extracted from current main:
https://github.com/PX4/PX4-Autopilot/blob/646f711b1a3b700bb55f3eec909ff53c728afe1f/src/modules/ekf2/EKF2.cpp#L2656

It might be because I'm not familiar enough with the codebase, but this could create an uneven load on the _work queues_. I think it would make more sense to try to assign each iteration to a different queue.

**Example**:
2 IMUs and 3 MAGs, something that could happen in a multi GPS situation.

This would generate something like this:

| EKF      | imu0 mag0 | imu1 mag0 | imu0 mag1 | imu1 mag1 | imu0 mag2 | imu1 mag2 |
|----------|-----------|-----------|-----------|-----------|-----------|-----------|
| Current  | wq0       | wq1       | wq0       | wq1       | wq0       | wq1       |
| Optimal? | wq0       | wq1       | wq2       | wq3       | wq0       | wq1       |

antonio-sc66 committed:
EKF2/Utility. Add sorting function

Signed-off-by: Antonio Sanjurjo C <74329840+antonio-sc66@users.noreply.github.com>

antonio-sc66 committed:
EKF2. Fix. Use existing sensor priority to spawn multi-ekf instances

- Previous version didn't work on most embedded targets.
- Sorting is now performed using the custom algorithm
- Variables are normal arrays instead of pairs

Signed-off-by: Antonio Sanjurjo C <74329840+antonio-sc66@users.noreply.github.com>

antonio-sc66 commented:
Sorting function would probably be better somewhere in src/lib.

Prints are still on for testing purposes, choose if this prints should be let on or be removed:

```
INFO  [ekf2] Initialized MAGs (positive priority): 2
INFO  [ekf2] MAGs priorities - i: 0  MAG1  Priority: 50
INFO  [ekf2] MAGs priorities - i: 1  MAG0  Priority: 50
INFO  [ekf2] MAGs priorities - i: 2  MAG3  Priority: -1
INFO  [ekf2] MAGs priorities - i: 3  MAG2  Priority: -1
INFO  [ekf2] Initialized IMUs (positive priority): 1
INFO  [ekf2] IMUs priorities - i: 0  ACC0  Priority: 50
INFO  [ekf2] IMUs priorities - i: 1  ACC3  Priority: -1
INFO  [ekf2] IMUs priorities - i: 2  ACC2  Priority: -1
INFO  [ekf2] IMUs priorities - i: 3  ACC1  Priority: -1
```

antonio-sc66 commented:
*diatone_mamba-f405-mk2* builds fine locally, with 99.89% of the flash region, the same as with current main:

```
make diatone_mamba-f405-mk2_default
-- PX4 version: v1.14.0-beta2-311-gded751e3ae (1.14.0)
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.8.10", minimum required is "3") 
-- PX4 config file: /home/user/user_PX4/boards/diatone/mamba-f405-mk2/default.px4board
-- PLATFORM nuttx
-- TOOLCHAIN arm-none-eabi
-- ARCHITECTURE cortex-m4
-- ROMFSROOT px4fmu_common
-- CONSTRAINED_FLASH y
-- NO_HELP y
-- CONSTRAINED_MEMORY y
-- EXTERNAL_METADATA y
-- SERIAL_URT6 /dev/ttyS2
-- SERIAL_TEL2 /dev/ttyS1
-- PX4 config: diatone_mamba-f405-mk2_default
-- PX4 platform: nuttx
-- The CXX compiler identification is GNU 9.3.1
-- The C compiler identification is GNU 9.3.1
-- The ASM compiler identification is GNU
-- Found assembler: /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc
-- Check for working CXX compiler: /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-g++
-- Check for working CXX compiler: /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc
-- Check for working C compiler: /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- cmake build type: MinSizeRel
-- ccache enabled (export CCACHE_DISABLE=1 to disable)
-- ROMFS: ROMFS/px4fmu_common
-- ROMFS:  Adding platforms/nuttx/init/stm32/rc.board_arch_defaults -> /etc/init.d/rc.board_arch_defaults
-- ROMFS:  Adding boards/diatone/mamba-f405-mk2/init/rc.board_defaults -> /etc/init.d/rc.board_defaults
-- ROMFS:  Adding boards/diatone/mamba-f405-mk2/init/rc.board_sensors -> /etc/init.d/rc.board_sensors
-- ROMFS:  Adding boards/diatone/mamba-f405-mk2/init/rc.board_extras -> /etc/init.d/rc.board_extras
-- Configuring done
-- Generating done
-- Build files have been written to: /home/user/user_PX4/build/diatone_mamba-f405-mk2_default
[0/733] git submodule src/drivers/gps/devices
[4/733] git submodule platforms/nuttx/NuttX/nuttx
[8/733] git submodule src/modules/mavlink/mavlink
[731/733] Linking CXX executable diatone_mamba-f405-mk2_default.elf
Memory region         Used Size  Region Size  %age Used
           flash:     1014649 B       992 KB     99.89%
            sram:       22612 B       128 KB     17.25%
          ccsram:          0 GB        64 KB      0.00%
[733/733] Creating /home/user/user_PX4/build/diatone_mamba-f405-mk2_default/diatone_mamba-f405-mk2_default.px4
```

antonio-sc66 commented:
**Mission** and **Offboard** failing tests don't have their sensors priorities set properly.

Example (build standard_vtol):
```
 [   8.311|px4       ] INFO  [ekf2] MAGs priorities - i: 0  MAG1  Priority: 50
[   8.312|px4       ] INFO  [ekf2] MAGs priorities - i: 1  MAG0  Priority: 50
[   8.312|px4       ] INFO  [ekf2] MAGs priorities - i: 2  MAG3  Priority: -1
[   8.312|px4       ] INFO  [ekf2] MAGs priorities - i: 3  MAG2  Priority: -1
[   8.312|px4       ] INFO  [ekf2] Initialized IMUs (positive priority): 0
[   8.312|px4       ] INFO  [ekf2] IMUs priorities - i: 0  ACC3  Priority: -1
[   8.312|px4       ] INFO  [ekf2] IMUs priorities - i: 1  ACC2  Priority: -1
[   8.312|px4       ] INFO  [ekf2] IMUs priorities - i: 2  ACC1  Priority: -1
[   8.312|px4       ] INFO  [ekf2] IMUs priorities - i: 3  ACC0  Priority: -1
[   8.312|px4       ] ERROR [ekf2] start failed (-1)
```

They fail as expected now, before this PR priorities weren't taken into consideration

bresch commented on src/modules/ekf2/EKF2.cpp:
```diff
@@ -2625,9 +2626,81 @@ int EKF2::task_spawn(int argc, char *argv[])
-		const int multi_instances = math::min(imu_instances * mag_instances, static_cast<int32_t>(EKF2_MAX_INSTANCES));
-		int multi_instances_allocated = 0;
+		// Number of instances is limited by the number of valid sensors (not uninitialized or disabled) that are beeing published
+		const uint8_t multi_instances = static_cast<uint8_t>(math::min(valid_imu_instances * mag_instances,
+						valid_imu_instances * static_cast<int32_t>(valid_mag_instances), static_cast<int32_t>(EKF2_MAX_INSTANCES)));
```
You also want to be limited by the param directly, right? We should then take the min of the param and the valid instances for both imu and mags.
```
		multi_instances = min(min(imu_instances, valid_imu_instances) * min(mag_instances, valid_mag_instances), EKF2_MAX_INSTANCES);
```

antonio-sc66 replied:
That's correct, I will modify this

bresch commented on src/modules/ekf2/EKF2.cpp:
```diff
@@ -2641,32 +2714,41 @@ int EKF2::task_spawn(int argc, char *argv[])
+			uint8_t selected_imu = 0;
 
-				for (uint8_t imu = 0; imu < imu_instances; imu++) {
+			for (uint8_t mag = 0; mag < valid_mag_instances; mag++) {
```
same as before, if `valid_mag_instances` > `mag_instances` you'll create more instances than required.

antonio-sc66 replied:
In this case, the limited number of instances can't be used instead of valid_*_instances as the sensor might not be advertising and multiple sensors might have to be checked in order to create a valid instance. Some checks have to be added

antonio-sc66 replied:
A return clause has been added to prevent excessive spawns

bresch reviewed (changes requested):
Nice! Looks overall good to me.

antonio-sc66 committed:
EKF2. Bugfix. Limit the number of multi-EKF sensor instances based on EKF2_MULTI_*** param.

- Fix. Correct the multi_instances variable calculation
- Perform early return once all the multi EKF instances have been created
- Rename some sensor related variables to match naming style

Signed-off-by: Antonio Sanjurjo C <74329840+antonio-sc66@users.noreply.github.com>

antonio-sc66 commented:
Hi @bresch and @dagar, can this MR be reviewed?

MaEtUgR commented:
Maintinaer call: The priorities are availble here: https://github.com/PX4/PX4-Autopilot/blob/5ca22df55c0016c9dd96b91139781206024aa0f8/msg/SensorsStatus.msg#L11
They need to be accounted for to configure the EKF instances.

antonio-sc66 committed:
EFK2. task_spawn. Use sensors_status instead of parameter values for sensor priorities

Signed-off-by: Antonio Sanjurjo C <74329840+antonio-sc66@users.noreply.github.com>

antonio-sc66 commented:
Hi @MaEtUgR, there's a weird behavior when I read the values from the topic vs when I read them from the params.

I've added a temporal print so that it can be compared:

```
INFO  [ekf2] MAGs priorities (from param) - MAG0  Priority: 50
INFO  [ekf2] MAGs priorities (from param) - MAG1  Priority: 50
INFO  [ekf2] MAGs priorities (from param) - MAG2  Priority: -1
INFO  [ekf2] MAGs priorities (from param) - MAG3  Priority: -1
INFO  [ekf2] IMUs priorities (from param) - IMU0  Priority: 50
INFO  [ekf2] IMUs priorities (from param) - IMU1  Priority: -1
INFO  [ekf2] IMUs priorities (from param) - IMU2  Priority: -1
INFO  [ekf2] IMUs priorities (from param) - IMU3  Priority: -1
INFO  [ekf2] Initialized MAGs (positive priority): 1
INFO  [ekf2] MAGs priorities - i: 0  MAG0  Priority: 50
INFO  [ekf2] MAGs priorities - i: 1  MAG3  Priority: 0
INFO  [ekf2] MAGs priorities - i: 2  MAG2  Priority: 0
INFO  [ekf2] MAGs priorities - i: 3  MAG1  Priority: 0
INFO  [ekf2] Initialized IMUs (positive priority): 1
INFO  [ekf2] IMUs priorities - i: 0  ACC0  Priority: 50
INFO  [ekf2] IMUs priorities - i: 1  ACC3  Priority: 0
INFO  [ekf2] IMUs priorities - i: 2  ACC2  Priority: 0
INFO  [ekf2] IMUs priorities - i: 3  ACC1  Priority: 0
``` 

When reading from parameters, there are two MAGs with positive priority: MAG0 and MAG1.
When reading from the topics only MAG0 has positive priority.

Is this behavior expected? It looks like it's due to some internal filtering.

Otherwise, everything seems to work as before with the revised implementation

antonio-sc66 commented:
Everything seems to be working fine with Gazebo.

New simulation execution output:

```
______  __   __    ___ 
| ___ \ \ \ / /   /   |
| |_/ /  \ V /   / /| |
|  __/   /   \  / /_| |
| |     / /^\ \ \___  |
\_|     \/   \/     |_/

px4 starting.

INFO  [px4] startup script: /bin/sh etc/init.d-posix/rcS 0
INFO  [init] found model autostart file as SYS_AUTOSTART=10015
INFO  [param] selected parameter default file parameters.bson
INFO  [param] importing from 'parameters.bson'
INFO  [parameters] BSON document size 404 bytes, decoded 404 bytes (INT32:14, FLOAT:6)
INFO  [param] selected parameter backup file parameters_backup.bson
INFO  [dataman] data manager file './dataman' size is 7868392 bytes
INFO  [init] PX4_SIM_HOSTNAME: localhost
INFO  [simulator_mavlink] Waiting for simulator to accept connection on TCP port 4560
INFO  [simulator_mavlink] Simulator connected on TCP port 4560.
INFO  [lockstep_scheduler] setting initial absolute time to 3132000 us
INFO  [commander] LED: open /dev/led0 failed (22)
WARN  [health_and_arming_checks] Preflight Fail: ekf2 missing data
INFO  [ekf2] Initialized MAGs (positive priority): 1
INFO  [ekf2] MAGs priorities - i: 0  MAG0  Priority: 50
INFO  [ekf2] MAGs priorities - i: 1  MAG3  Priority: 0
INFO  [ekf2] MAGs priorities - i: 2  MAG2  Priority: 0
INFO  [ekf2] MAGs priorities - i: 3  MAG1  Priority: 0
INFO  [ekf2] Initialized IMUs (positive priority): 1
INFO  [ekf2] IMUs priorities - i: 0  ACC0  Priority: 50
INFO  [ekf2] IMUs priorities - i: 1  ACC3  Priority: 0
INFO  [ekf2] IMUs priorities - i: 2  ACC2  Priority: 0
INFO  [ekf2] IMUs priorities - i: 3  ACC1  Priority: 0
INFO  [mavlink] mode: Normal, data rate: 4000000 B/s on udp port 18570 remote port 14550
Gazebo multi-robot simulator, version 11.8.1
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.1.42
INFO  [mavlink] mode: Onboard, data rate: 4000000 B/s on udp port 14580 remote port 14540
INFO  [mavlink] mode: Onboard, data rate: 4000 B/s on udp port 14280 remote port 14030
INFO  [mavlink] mode: Gimbal, data rate: 400000 B/s on udp port 13030 remote port 13280
INFO  [logger] logger started (mode=all)
INFO  [logger] Start file log (type: full)
INFO  [logger] [logger] ./log/2023-12-11/22_54_56.ulg	
INFO  [logger] Opened full log file: ./log/2023-12-11/22_54_56.ulg
INFO  [mavlink] MAVLink only on localhost (set param MAV_{i}_BROADCAST = 1 to enable network)
INFO  [mavlink] MAVLink only on localhost (set param MAV_{i}_BROADCAST = 1 to enable network)
INFO  [px4] Startup script returned successfully
pxh> INFO  [tone_alarm] home set
INFO  [mavlink] partner IP: 127.0.0.1

pxh> commaINFO  [commander] Ready for takeoff!
pxh> commander takeoff
pxh> INFO  [commander] Armed by internal command	
INFO  [tone_alarm] arming warning
INFO  [navigator] Using default takeoff altitude: 2.5 m	
INFO  [commander] Takeoff detected	

pxh> commander land
pxh> INFO  [commander] Landing at current position	
INFO  [commander] Landing detected	
INFO  [commander] Disarmed by landing	
INFO  [tone_alarm] notify neutral
INFO  [logger] closed logfile, bytes written: 8026247

pxh> shutdown
Exiting NOW.
```

bresch commented:
@dagar any comments on this?

AlexKlimaj commented:
@dagar I am also interested in this.
